src/metric.rs

```diff
@@ -0,0 +1,305 @@
+    fn type_code(&self) -> u32 {
+        STRING_METRIC_TYPE_CODE
+    }
+
+    fn write_to_writer<W: WriteBytesExt>(&self, mut writer: &mut W) -> io::Result<()> {
+        writer.write_all(CString::new(self.as_str())?.to_bytes_with_nul())
+    }
+}
+
+#[derive(Copy, Clone)]
+pub enum SpaceScale {
+    /// byte
+    Byte = 0,
+    /// kilobyte (1024) 
+    KByte,
+    /// megabyte (1024^2)
+    MByte,
+    /// gigabyte (1024^3)
+    GByte,
+    /// terabyte (1024^4)
+    TByte,
+    /// petabyte (1024^5)
+    PByte,
+    /// exabyte (1024^6)
+    EByte
+}
+
+#[derive(Copy, Clone)]
+pub enum TimeScale {
+    /// nanosecond
+    NSec = 0,
+    /// microsecond
+    USec,
+    /// millisecond
+    MSec,
+    /// second
+    Sec,
+    /// minute
+    Min,
+    /// hour
+    Hour
+}
+
+#[derive(Copy, Clone)]
+pub enum CountScale {
+    One = 0
+}
+
+
+pub struct Unit {
+    pub space_scale: SpaceScale,
+    pub time_scale: TimeScale,
+    pub count_scale: CountScale,
+    pub space_dim: i8,
+    pub time_dim: i8,
+    pub count_dim: i8
+}
+
+impl Unit {
+    pub fn empty() -> Self {
+        Unit {
+            space_scale: SpaceScale::Byte,
+            time_scale: TimeScale::NSec,
+            count_scale: CountScale::One,
+            space_dim: 0,
+            time_dim: 0,
+            count_dim: 0
+        }
+    }
+
+    pub fn space(space_scale: SpaceScale) -> Self {
+        let mut unit = Self::empty();
+        unit.space_scale = space_scale;
+        unit.space_dim = 1;
+        unit
+    }
+
+    pub fn time(time_scale: TimeScale) -> Self {
+        let mut unit = Self::empty();
+        unit.time_scale = time_scale;
+        unit.time_dim = 1;
+        unit
+    }
+
+    pub fn count() -> Self {
+        let mut unit = Self::empty();
+        unit.count_dim = 1;
+        unit
+    }
+
+    /*
+        pmapi representation of a unit. below, 31 refers to MSB
+        bits 31 - 28 : space dim   (signed)
+             27 - 24 : time dim    (signed)
+             23 - 20 : count dim   (signed)
+             19 - 16 : space scale (unsigned)
+             15 - 12 : time scale  (unsigned)
+             11 - 8  : count scale (unsigned)
+              7 - 0  : zero pad
+    */
+    fn pmapi_repr(&self) -> u32 {
+        let mut repr = 0;
+
+        repr |= ((self.space_dim as i32) & ((1 << 4) - 1)) << 28;
+        repr |= ((self.time_dim  as i32) & ((1 << 4) - 1)) << 24;
+        repr |= ((self.count_dim as i32) & ((1 << 4) - 1)) << 20;
+
+        repr |= (self.space_scale as i32) << 16;
+        repr |= (self.time_scale  as i32) << 12;
+        repr |= (self.count_scale as i32) << 8;
+
+        repr as u32
+    }
+}
+
+#[derive(Copy, Clone)]
+pub enum Semamtics {
+    Counter  = 1,
+    Instant  = 3,
+    Discrete = 4
+}
+
+pub struct Metric<T> {
+    name: String,
+    item: u32,
+    sem: Semamtics,
+    indom: u32,
+    unit: u32,
+    shorthelp: String,
+    longhelp: String,
+    val: T,
+    mmap_view: MmapViewSync
+}
```

Comment:
That is a problem, actually that's the thing, and I think @natoscott can attest to this, when we designed speed, we designed interfaces and structs such that apart from PCP, the interfaces should be agnostic enough so that the entire library can provide an API for another system. Having Metric, Client etc. as traits helps. Also, we should probably go for static dispatch instead of dynamic, as it is not required, so the export function can have a generic parameter `M: Metric`.